modules/apps/portal-search-elasticsearch7/portal-search-elasticsearch7-impl/src/main/java/com/liferay/portal/search/elasticsearch7/internal/ElasticsearchIndexSearcher.java

```diff
@@ -413,6 +415,82 @@ private String[] _getIndexes(
 		return new String[] {indexName};
 	}
 
+	private SearchHit _getLastSearchHit(
```

Comment:
Looks much cleaner! I think it could still be simplified and made a bit easier to read. How about extracting the _searchWithDeepPagination () code further, to something like this (please feel free to improve):
```

	private SearchHit _getLastSearchHit(
		int maxResultWindow, SearchSearchRequest searchSearchRequest,
		int start) {

		int documentsToSkip = 0;

		if (start < maxResultWindow) {
			searchSearchRequest.setStart(start - 1);
		}
		else {
			searchSearchRequest.setStart(maxResultWindow - 1);

			documentsToSkip = start % maxResultWindow;
		}

		searchSearchRequest.setSize(1);

		SearchSearchResponse searchSearchResponse =
			_searchEngineAdapter.execute(searchSearchRequest);

		SearchHit lastSearchHit = _getLastSearchHit(searchSearchResponse);

		if (lastSearchHit == null) {
			return null;
		}

		int maxResultWindowPages = start / maxResultWindow;

		for (int i = 1; i < maxResultWindowPages; i++) {
			lastSearchHit = _getLastSearchHit(
				lastSearchHit.getSortValues(), searchSearchRequest,
				maxResultWindow, 0);

			if (lastSearchHit == null) {
				return null;
			}
		}

		if (documentsToSkip > 0) {
			lastSearchHit = _getLastSearchHit(
				lastSearchHit.getSortValues(), searchSearchRequest,
				documentsToSkip, 0);

			if (lastSearchHit == null) {
				return null;
			}
		}

		return lastSearchHit;
	}

	private SearchHit _getLastSearchHit(
		Object[] searchAfter, SearchSearchRequest searchSearchRequest, int size,
		int start) {

		if (searchAfter != null) {
			searchSearchRequest.setSearchAfter(searchAfter);
		}

		searchSearchRequest.setSize(size);
		searchSearchRequest.setStart(start);

		SearchSearchResponse searchSearchResponse =
			_searchEngineAdapter.execute(searchSearchRequest);

		return _getLastSearchHit(searchSearchResponse);
	}

	private SearchHit _getLastSearchHit(
		SearchSearchResponse searchSearchResponse) {

		SearchHits searchHits = searchSearchResponse.getSearchHits();

		List<SearchHit> searchHitList = searchHits.getSearchHits();

		if (searchHitList.isEmpty()) {
			return null;
		}

		return searchHitList.get(searchHitList.size() - 1);
	}

	private Hits _searchWithDeepPagination(
		int end, Query query, SearchContext searchContext,
		SearchRequest searchRequest,
		SearchResponseBuilder searchResponseBuilder, int start) {

		SearchSearchRequest searchSearchRequest =
			_createSearchSearchRequestWithDeepPagination(
				searchRequest, searchContext, query);

		SearchSearchResponse searchSearchResponse = null;

		int maxResultWindow = 10000; // GL: should be configurable

		try {
			if (end > maxResultWindow) {
				SearchHit lastSearchHit = _getLastSearchHit(
					start, searchSearchRequest, maxResultWindow);

				if (lastSearchHit == null) {
					return new HitsImpl();
				}

				searchSearchRequest.setSearchAfter(
					lastSearchHit.getSortValues());

				searchSearchRequest.setStart(0);
			}
			else {
				searchSearchRequest.setStart(start);
			}

			searchSearchRequest.setSize(end - start);

			searchSearchResponse = _searchEngineAdapter.execute(
				searchSearchRequest);
		}
		catch (RuntimeException runtimeException) {
			throw runtimeException;
		}
		finally {
			PointInTime pointInTime = searchSearchRequest.getPointInTime();

			_searchEngineAdapter.execute(
				new ClosePointInTimeRequest(pointInTime.getPointInTimeId()));
		}

		_populateResponse(searchSearchResponse, searchResponseBuilder);

		searchResponseBuilder.searchHits(searchSearchResponse.getSearchHits());

		return searchSearchResponse.getHits();
	}

```